### ISSUE TITLE
False positive error for comparing list of literal strings vs list of strings

### ISSUE THREAD

**Issue by RobertCraigie** [microsoft/pyright#6099]
**Describe the bug**

This code produces the following error:

```python
VERSION = '1.2.3'

if VERSION.split('.') > ['1', '2']:
    ...
```

```
example.py:3:4 - error: Operator ">" not supported for types "list[LiteralString]" and "list[str]" (reportGeneralTypeIssues)
```

This code passed type checks prior to `v1.1.329` and works at runtime.

**VS Code extension or command-line**

CLI, v1.1.330

**Comment by erictraut** [comment]
Yes, I think the new behavior is correct from a type perspective, although I agree that it's annoying and counterintuitive. It's the consequence of a recent bug fix.

In your code sample, the type of the expression `VERSION.split('.')` is `list[LiteralString]`, and the type of the expression `['1', '2']` is `list[str]`. These types are not comparable according to `list.__gt__`.

I don't see a way to "fix" this in a type checker, since any change would re-introduce the bug that was recently fixed. I also don't see a straightforward way to fix this in typeshed.

Here are a couple of workarounds:
```python
from typing import LiteralString

VERSION = "1.2.3"
min_version: list[LiteralString] = ["1", "2"]

if VERSION.split(".") > min_version:
    ...
```

```python
VERSION = "1.2.3"

if tuple(VERSION.split(".")) > ("1", "2"):
    ...
```

**Comment by kwshi** [comment]
@erictraut Can you briefly explain the difference, or refer me to some discussion about why this "new behavior" was introduced, and why it makes more sense? I'm running into a similar issue-- I'm guessing it's the same cause?
```python
s = 'a, b, c'.split(', ')
def f(x: str):
  return s.index(x)
```

**Comment by erictraut** [comment]
@kwshi, this is related to the concept of "invariance". The `list` type is generic, and its type parameter is invariant. That means `list[str]` is not a subtype of `list[str | None]` even though `str` is a subtype of `str | None`. Likewise, it means `list[LiteralString]` is not a subtype of `list[str]` even though `LiteralString` is a subtype of `str`.

In your example, the type of `s` is `list[LiteralString]`, and the `list.index` method requires that the value passed to it is compatible with the list's type argument. You're passing a `str` argument, but a `LiteralString` is required in this case. 

This isn't a case of "what behavior makes more sense". These rules are dictated by the type system, and it's the job of a type checker to enforce these rules. A type checker can't simply choose to ignore certain type compatibility rules. The "old behavior" in this case was a bug.

I think it's unfortunate that `str.split` is defined in the way it is. I've filed [this bug](https://github.com/python/typeshed/issues/10887) in the typeshed project. Feel free to add to that discussion if you'd like.

**Comment by RobertCraigie** [comment]
@erictraut would it be possible to update the type inferrer to handle the in-line expression case? i.e. in a comparison expression like this, if there are multiple choices for the broadness of the type thats inferred for the RHS, then whichever type is assignable to the LHS would be chosen.

e.g. in my case I think it would be reasonable behaviour for the expression `['1', '2']` to result in `list[LiteralString]` because it's defined in-line.

**Comment by erictraut** [comment]
I'm not sure what you mean by "in-line expression case". Are you suggesting that pyright change its inference rules specifically in the context of the RHS of a comparison operation? No, that wouldn't be a good idea. That would create a different set of false positives, and it would be difficult to justify and explain. Inference rules need to be consistent.

One of the challenges specifically with comparison operators (and other binary operators) is that they are not simple calls to magic methods in Python. They're complex operations that involve potentially multiple magic method calls (e.g. `__add__` and `__radd__`). For that reason, type checkers can't use bidirectional type inference for binary operators. Normal type inference rules apply here.

You can override the default inference behaviors yourself through the use of a temporary variable as I suggested above.

**Comment by RobertCraigie** [comment]
I see, my thinking was that it's essentially the same situation as this:
```py
def foo(data: list[str]) -> None:
    ...

def bar(data: list[LiteralString]) -> None:
    ...

foo(['a', 'b', 'c'])
bar(['a', 'b', 'c'])
```

Where pyright allows the argument expression to be applied to both `list[str]` and `list[LiteralString]`. I was not aware that handling binary operations was so much more complicated.
